ClassNotFoundException when running JettyLauncher with Java 9

GWT version: 2.8.2
JDK: 1.9

---

##### Description
When running GWTUnitTests with Java9 in ProdMode the following Error occures in log:

```
[ERROR] jreLeakPrevention.gcDaemonFail
java.lang.ClassNotFoundException: sun.misc.GC
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:582)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:185)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:496)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:292)
	at com.google.gwt.dev.shell.jetty.JettyLauncher.jreLeakPrevention(JettyLauncher.java:899)
	at com.google.gwt.dev.shell.jetty.JettyLauncher.start(JettyLauncher.java:722)
	at com.google.gwt.dev.DevMode.doStartUpServer(DevMode.java:636)
	at com.google.gwt.dev.DevModeBase.startUp(DevModeBase.java:923)
	at com.google.gwt.junit.JUnitShell.getUnitTestShell(JUnitShell.java:708)
	at com.google.gwt.junit.JUnitShell.runTest(JUnitShell.java:682)
	at com.google.gwt.junit.client.GWTTestCase.runTest(GWTTestCase.java:421)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at com.google.gwt.junit.client.GWTTestCase.run(GWTTestCase.java:247)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:369)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:275)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:239)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:160)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:373)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:334)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:119)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:407)
```

See also: https://travis-ci.org/TDesjardins/gwt-ol/jobs/291208146#L1547

The cause is presumably that a class (sun.misc.GC) has been moved in Java 9:

https://bugs.openjdk.java.net/browse/JDK-8153286

which is used in the JettyLauncher:

https://github.com/gwtproject/gwt/blob/a90832fec9a56f4cb83ed009c7048e0a8d196fe8/dev/core/src/com/google/gwt/dev/shell/jetty/JettyLauncher.java#L899

##### Steps to reproduce

https://travis-ci.org/TDesjardins/gwt-ol/jobs/291208146

##### Known workarounds

Use Java 8

--- a/dev/core/src/com/google/gwt/dev/shell/jetty/JettyLauncher.java
+++ b/dev/core/src/com/google/gwt/dev/shell/jetty/JettyLauncher.java
@@ -42,6 +42,10 @@ import org.eclipse.jetty.servlet.ServletContextHandler;
 import org.eclipse.jetty.util.component.AbstractLifeCycle;
 import org.eclipse.jetty.util.log.Log;
 import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.util.preventers.AppContextLeakPreventer;
+import org.eclipse.jetty.util.preventers.DOMLeakPreventer;
+import org.eclipse.jetty.util.preventers.GCThreadLeakPreventer;
+import org.eclipse.jetty.util.preventers.SecurityProviderLeakPreventer;
 import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.eclipse.jetty.webapp.ClasspathPattern;
 import org.eclipse.jetty.webapp.Configuration;
@@ -59,10 +63,6 @@ import java.util.Collections;
 import java.util.Enumeration;
 import java.util.Iterator;
 import java.util.List;
-
-import javax.imageio.ImageIO;
-import javax.xml.parsers.DocumentBuilderFactory;
-import javax.xml.parsers.ParserConfigurationException;
 
 /**
  * A {@link ServletContainerLauncher} for an embedded Jetty server.
@@ -744,6 +744,7 @@ public class JettyLauncher extends ServletContainerLauncher {
     ServerConnector connector = getConnector(server, logger);
     setupConnector(connector, bindAddress, port);
     server.addConnector(connector);
+    addPreventers(server);
 
     Configuration.ClassList cl = Configuration.ClassList.setServerDefault(server);
     try {
@@ -859,6 +860,37 @@ public class JettyLauncher extends ServletContainerLauncher {
      return config;
   }
 
+  private void addPreventers(Server server) {
+    // Trigger a call to sun.awt.AppContext.getAppContext(). This will
+    // pin the common class loader in memory but that shouldn't be an
+    // issue.
+    server.addBean(new AppContextLeakPreventer());
+
+    /*
+     * Several components end up calling: sun.misc.GC.requestLatency(long)
+     *
+     * Those libraries / components known to trigger memory leaks due to 
+     * eventual calls to requestLatency(long) are:
+     * - javax.management.remote.rmi.RMIConnectorServer.start()
+     */
+    server.addBean(new GCThreadLeakPreventer());
+
+    /*
+     * Creating a MessageDigest during web application startup initializes the 
+     * Java Cryptography Architecture. Under certain conditions this starts a 
+     * Token poller thread with TCCL equal to the web application class loader.
+     *
+     * Instead we initialize JCA right now.
+     */
+    server.addBean(new SecurityProviderLeakPreventer());
+
+    /*
+     * Haven't got to the root of what is going on with this leak but if a web app is the first to
+     * make the calls below the web application class loader will be pinned in memory.
+     */
+    server.addBean(new DOMLeakPreventer());
+  }
+
   private void checkStartParams(TreeLogger logger, int port, File appRootDir) {
     if (logger == null) {
       throw new NullPointerException("logger cannot be null");
@@ -899,37 +931,6 @@ public class JettyLauncher extends ServletContainerLauncher {
    * (http://www.apache.org/).
    */
   private void jreLeakPrevention(TreeLogger logger) {
-    // Trigger a call to sun.awt.AppContext.getAppContext(). This will
-    // pin the common class loader in memory but that shouldn't be an
-    // issue.
-    ImageIO.getCacheDirectory();
-
-    /*
-     * Several components end up calling: sun.misc.GC.requestLatency(long)
-     *
-     * Those libraries / components known to trigger memory leaks due to
-     * eventual calls to requestLatency(long) are: -
-     * javax.management.remote.rmi.RMIConnectorServer.start()
-     */
-    try {
-      Class<?> clazz = Class.forName("sun.misc.GC");
-      Method method = clazz.getDeclaredMethod("requestLatency",
-          new Class[]{long.class});
-      method.invoke(null, Long.valueOf(3600000));
-    } catch (ClassNotFoundException e) {
-      logger.log(TreeLogger.ERROR, "jreLeakPrevention.gcDaemonFail", e);
-    } catch (SecurityException e) {
-      logger.log(TreeLogger.ERROR, "jreLeakPrevention.gcDaemonFail", e);
-    } catch (NoSuchMethodException e) {
-      logger.log(TreeLogger.ERROR, "jreLeakPrevention.gcDaemonFail", e);
-    } catch (IllegalArgumentException e) {
-      logger.log(TreeLogger.ERROR, "jreLeakPrevention.gcDaemonFail", e);
-    } catch (IllegalAccessException e) {
-      logger.log(TreeLogger.ERROR, "jreLeakPrevention.gcDaemonFail", e);
-    } catch (InvocationTargetException e) {
-      logger.log(TreeLogger.ERROR, "jreLeakPrevention.gcDaemonFail", e);
-    }
-
     /*
      * Calling getPolicy retains a static reference to the context class loader.
      */
@@ -954,15 +955,6 @@ public class JettyLauncher extends ServletContainerLauncher {
     }
 
     /*
-     * Creating a MessageDigest during web application startup initializes the
-     * Java Cryptography Architecture. Under certain conditions this starts a
-     * Token poller thread with TCCL equal to the web application class loader.
-     *
-     * Instead we initialize JCA right now.
-     */
-    java.security.Security.getProviders();
-
-    /*
      * Several components end up opening JarURLConnections without first
      * disabling caching. This effectively locks the file. Whilst more
      * noticeable and harder to ignore on Windows, it affects all operating
@@ -983,18 +975,6 @@ public class JettyLauncher extends ServletContainerLauncher {
       logger.log(TreeLogger.ERROR, "jreLeakPrevention.jarUrlConnCacheFail", e);
     } catch (IOException e) {
       logger.log(TreeLogger.ERROR, "jreLeakPrevention.jarUrlConnCacheFail", e);
-    }
-
-    /*
-     * Haven't got to the root of what is going on with this leak but if a web
-     * app is the first to make the calls below the web application class loader
-     * will be pinned in memory.
-     */
-    DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
-    try {
-      factory.newDocumentBuilder();
-    } catch (ParserConfigurationException e) {
-      logger.log(TreeLogger.ERROR, "jreLeakPrevention.xmlParseFail", e);
     }
   }
 }

